PocketShop/Views/CustomerViews/CouponListView.swift

```diff
@@ -0,0 +1,76 @@
+import SwiftUI
+
+struct CouponListView: View {
+    @Environment(\.presentationMode) var presentationMode
+    @ObservedObject var cartViewModel: CustomerCartScreen.ViewModel
+
+    var body: some View {
+        NavigationView {
+            Group {
+                if cartViewModel.customerCoupons.isEmpty {
+                    Text("You don't have any coupons right now")
+                        .font(.title)
+                } else {
+                    ScrollView(.vertical) {
+                        VStack(alignment: .leading) {
+                            ForEach(cartViewModel.customerCoupons, id: \.self) { coupon in
+                                couponItem(coupon: coupon, quantity: cartViewModel.customerCouponCount(coupon))
+                                    .opacity(cartViewModel.canApplyCoupon(coupon) ? 1.0 : 0.5)
+                                Divider()
+                            }
+                        }
+                    }
+                    .frame(maxWidth: .infinity, alignment: .leading)
+                }
+            }
+            .padding()
+            .navigationTitle("Your Coupons")
+        }
+    }
+
+    @ViewBuilder
+    func couponItem(coupon: Coupon, quantity: Int) -> some View {
+        HStack {
+            VStack(alignment: .leading) {
+                Text("\(coupon.description)")
+                    .font(.appHeadline)
+
+                getDiscountText(coupon: coupon)
+
+                Text("Minimum order: $\(String(format: "%.2f", coupon.minimumOrder))")
+
+                Text("Coupons left: \(quantity)")
+                    .font(.subheadline)
+                    .padding(.top)
+            }
+
+            Spacer()
+
+            PSButton(title: cartViewModel.selectedCoupon == coupon ? "Remove Coupon" : "Apply Coupon") {
+                if cartViewModel.selectedCoupon == coupon {
+                    cartViewModel.deselectCoupon()
+                    presentationMode.wrappedValue.dismiss()
+                } else if cartViewModel.canApplyCoupon(coupon) {
+                    cartViewModel.applyCoupon(coupon: coupon)
+                    presentationMode.wrappedValue.dismiss()
+                }
+            }
+        }
+        .padding()
+    }
+
+    func getDiscountText(coupon: Coupon) -> Text {
```

Comment:
This function seems reused in `CouponRedemptionScreen` as well. Maybe place it in a ViewModel instead?

Reply:
I think ViewModel should not be constructing View. I prefer to keep the ViewModel just handling the logic and updates the view. It would be seem a bit strange if in the `CustomerViewModel`, there is one function that creates view and it's only used for the Coupon-related views. Let me know your thoughts on this!

Reply:
Ohh I see what you mean. Hmm in that case will making a View by itself work? Since views are globally accessible and can be reused in any other Views. It's kind of an alternative to the viewbuilders:
```swift
struct DiscountText: View {
    var coupon: Coupon
    var body: some View {
        ...
```

Reply:
Good idea! Will add that